code/server/server.cpp

```diff
@@ -0,0 +1,178 @@
+		return true;
+	}else{
+		return false;
+	}
+}
+
+int FIFO::dequeue()		//dequeue the first element of FIFO
+{
+	if(intQueLen>0){
+		intQueLen--;
+		if(intHead+1 < FIFO_SIZE){
+			intHead++;
+		}else{
+			intHead = 0;
+		}
+		FIFO_copy();	//copy the content of FIFO to global variable for monitoring
+		return true;
+	}else{
+		return false;
+	}
+}
+
+int FIFO::head_value()	//return the value of the first element of FIFO
+{
+	return (intFIFO[intHead]);
+}
+
+void FIFO::delete_element(int intNatural)	//delete the element of FIFO corespoinding to the argument "intNatural"
+{
+	int intLen = intQueLen;
+	for(int i=0; i<intLen; i++){
+		if(intNatural == intFIFO[intHead]){
+			dequeue();
+		}
+		else{
+			enqueue(intFIFO[intHead]);
+			dequeue();
+		}
+	}
+}
+
+void FIFO::FIFO_copy()
+{
+	if(intQueLen>0){
+		for(int i = 0; i < intQueLen; i++){
+			intFIFOtmp[i] = intFIFO[i];
+		}
+		intFIFOnum = intQueLen;
+	}
+}
+
+class Thread1 : public RateThread{
+public:
+	Thread1(int r):RateThread(r){}
+	virtual bool threadInit()
+	{
+		cout << "Starting thread" << endl;
+		return true;
+	}
+
+	virtual void afterStart(bool s)
+	{
+		if(s){
+			cout << "Thread started successfully" << endl;
+		}else{
+			cout << "Thread didn't start" << endl;
+		}
+	}
+
+	virtual void run()
+	{
+   		Semaphore Sema;
+   		Sema.wait();	//wait update procedure in main function
+
+   		//if FIFO has elements, show them
+		if (intFIFOnum > 0){
+			for (int i = 0; i < intFIFOnum; i++){
+				cout << "FIFO[" << i << "]=" << intFIFOtmp[i] << ",";
+			}
+			cout << endl;
+		}
+	}
+
+	virtual void threadRelease()
+	{
+		cout << "Goodbye from thread" << endl;
+	}
+};
+
+int main(int argc, char * argv[])
+{
+    Network yarp;	//initialize yarp
+    Port portP;
+    portP.open("/server");
+    FIFO Fifo;
+    Semaphore Sema;
+
+    //initialize grobal variables for monitoring of the content of FIFO
+    intFIFOnum = 0;
+    for(int i = 0; i<FIFO_SIZE; i++){
+    	intFIFOtmp[i] = 0;
+    }
+    
+    Thread1 t1(5000);	//report the content of FIFO every 5s
+    bool ok = t1.start();
+    if(!ok){
+    	cout << "The thread failed to initialize" << endl;
+    	return 0;
+    }
+
+	int intCNT = 1;		//set initial value of the counter
+
+	while(true){
+		Bottle botRequest;		//container for request from client
+		Bottle botCommand;		//container for command to client
+		portP.read(botRequest, true);		//wait to receive a request from client
+
+		if(botRequest.get(1).asInt() != 0){						//first connection from client sends 0, so skip
+			Fifo.delete_element(botRequest.get(1).asInt());		//delete the received element from FIFO
+		}
+		
+		//increment and push back the natural N
+		intCNT++;
+		Fifo.enqueue(intCNT);
+
+		if (Sema.check() == true){
+			Sema.post();
+		}
+		
+		//make a command and send it
+		botCommand.addString("vocab_item");
+		botCommand.addInt(intCNT);
+		botCommand.addInt(Fifo.head_value()-1);
+		portP.reply(botCommand);
+	}
+
+	return 0;
+}
```

Comment:
Yes you don't actually need to use `RateThread` if you use `RFModule` in this case, you can use the `RFModule::updateModule()` method to print the FIFO (instead of doing it in the `RateThread::run()` method).